sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3198,9 +3199,66 @@ class Dataset[T] private[sql](
   }
 
   /**
-   * Collect a Dataset as Arrow batches and serve stream to PySpark.
+   * Collect a Dataset as Arrow batches and serve stream to SparkR. It sends
+   * arrow batches in an ordered manner with buffering. This is inevitable
+   * due to missing R API that reads batches from socket directly. See ARROW-4512.
+   * Eventually, this code should be deduplicated by `collectAsArrowToPython`.
    */
-  private[sql] def collectAsArrowToPython(): Array[Any] = {
+  private[sql] def collectAsArrowToR(): Array[Any] = {
+    val timeZoneId = sparkSession.sessionState.conf.sessionLocalTimeZone
+
+    withAction("collectAsArrowToR", queryExecution) { plan =>
+      RRDD.serveToStream("serve-Arrow") { outputStream =>
```

Comment:
These codes were restored from previous Python side's Arrow optimization (see https://github.com/apache/spark/commit/ecaa495b1fe532c36e952ccac42f4715809476af). For now, due to the limitation of R API (see ARROW-3204), it should send Arrow batches at once (it cannot read batch by batch like Python side).

Once this is fixed, we can deduplicate the codes with `collectAsArrowToPython`.

Reply:
cc @BryanCutler

Reply:
> it should send Arrow batches at once (it cannot read batch by batch like Python side).

If I understand correctly, this means the arrow R api will create a data frame directly from a stream, and there is currently no way to get an array of record batches so they can be re-ordered?

Reply:
Yes, kind of. To be more explicit, I am buffering the batches in JVM side in an order by the reasons below:

1.. R API does not look supporting to read Arrow data from socket (ARROW-4512).
For instance, we cannot `read_arrow(socketConn)`

```
JVM side                                              R side    
[Arrow Batch][Arrow Batch][Arrow Batch] ---socket---> [Arrow Batch]...
```

Here, R side cannot read batch by batch. It is unable to only read batch's message by partially reading the batch.
There looks no way to know the batch's length in R side.


```
JVM side                                              R side    
[Arrow Stream Format (Arrow Batches)]   ---socket---> [Arrow Stream Format (Arrow Batches)]
```

Here is the current approach. I fully read the binary, and then read it via R API (`read_arrow(readRaw(socketConn))`)


2.. Yes, I also actually failed to find a way to construct Arrow table from Arrow batches in stream format via R APIs. However, I was thinking it's already blocked by the reason above so didn't investigate and file an issue at Arrow side.

3.. Since I need to buffer batches in JVM anyway, I wanted to get rid of batch order handling in R side.

Reply:
I will fix it as soon as ARROW-4512 at Arrow 0.14.0 is fixed anyway. I could elegantly deduplicate all of them once that's resolved IMHO.

Reply:
> For instance, we cannot `read_arrow(socketConn)`

If the Arrow R api follows C++ then there could be a layer of abstraction to create the reader, e.g. wrap the socket as an [InputStream](https://github.com/apache/arrow/blob/apache-arrow-0.12.0/r/R/io.R#L148)

This is what I've done in C++, sorry if I knew R a bit better I could tell you for sure :(
```c++
// Class to wrap a socket as a readable Arrow InputStream
class ArrowStreamClient : public arrow::io::InputStream {
 public:
  ArrowStreamClient(const std::string& host);
  ~ArrowStreamClient() override;

  arrow::Status Connect();
  arrow::Status Close() override;
  arrow::Status Tell(int64_t* position) const override;
  arrow::Status Read(int64_t nbytes, int64_t* bytes_read, void* out) override;
  arrow::Status Read(int64_t nbytes,
                     std::shared_ptr<arrow::Buffer>* out) override;

 private:
  const std::string host_;
  int sock_;
  int64_t pos_;
};

// Then to use it
auto in_stream_ =  std::make_shared<ArrowStreamClient>(dataset()->host_);
socket_stream->Connect());
arrow::ipc::RecordBatchStreamReader::Open(in_stream_.get(), &reader_));
reader_->ReadNext(&current_batch_));
```

Reply:
Yes, I 100% agree about the idea InputStream. I also tried to implement that but then realised that I should better avoid to hack it ...

Reply:
Yeah, it's a bit tricky to implement so would be great to have build into Arrow anyway